ADR: Add New repository library design (#1693)
theupdateframework/tuf: docs/repository-library-design.md

# Python-tuf repository API proposal: _minimal repository abstraction_

This is an attachment to ADR 10: _Repository library design built on top of

> Metadata API_, and documents the design proposal in Dec 2020.

Reviewer: ```suggestion
Metadata API_, and documents the design proposal in Dec 2021.
```
?

2. Provide canonical solutions for the difficult repository problems but avoid
   making implementation decisions
3. Keep python-tuf maintenance burden in mind: less is more

Why does this design look so different from both legacy python-tuf code and
other implementations?
* Most existing implementations are focused on a specific use case (typically a
  command line application): this is a valid design choice but severely limits
  goal #1
* The problem space contains many application decisions. Many implementations
  solve this by creating functions with 15 arguments: this design tries to find
  another way (#2)
* The Metadata API makes modifying individual pieces of metadata simpler. This,
  combined with good repository API design, should enable more variance in
  where things are implemented: The repository library does not have to
  implement every little detail as we can safely let specific implementations
  handle things, see goal #3
* This variance means we can start by implementing a minimal design: as
  experience from implementations is collected, we can then move implementation
  details into the library (goals #2, #3)

## Design

> ![Design: Application and library components](repository-library-design-ownership.jpg)

Reviewer: I was a bit confused by the word "generic" in "Generic App Code". Isn't this actually the code *specific* to an application? Maybe it's me, but I briefly thought that it means that it is something provided by python-tuf (generic and thus eligible for all applications).

Would "App management code" be a good choice of words?

(Same goes for the second diagram)

Author: yeah, "generic" was meant in the sense that _"this code does not need to be part of the class implementing Repository interface"_...

The design expects a fully functional repository application to contain code at
three levels:
* Repository library (abstract classes that are part of python-tuf)

>   * The Repository abstract class provides an ergonomic metadata editing API
>     for all code levels to use. It also implements some core edit actions like
>     snapshot update

Reviewer: I think it is important to be clear about what python-tuf actually implements and for what it provides abstract interfaces, e.g.

```suggestion
  * The Repository abstract class provides an ergonomic abstract metadata editing API
    for all code levels to use. It also provides concrete implementations for some core edit actions like
    snapshot update
```

  * A small amount of related functionality is also provided (private key
    management API, maybe repository validation)
  * is a very small library: possibly a few hundred lines of code
* Concrete Repository implementation (typically part of application code,
  implements interfaces provided by the repository API in python-tuf)
  * Contains the “application level” decisions that the Repository abstraction
    requires to operate: examples of application decisions include

>     * _when should “targets” metadata next expire when it is edited?_

Reviewer: nit: Case consistency:
```suggestion
    * _When should “targets” metadata next expire when it is edited?_
```

    * _What is the current “targets” metadata version? Where do we load it 
      from?_
    * _Where to store current “targets” after editing? Should the previous
      version be deleted from storage?_
* Actual application
  * Uses the Repository API to do the repository actions it needs to do

For context here’s a trivial example showing what “ergonomic editing” means --

> this key-adding code could be in the application or in the python-tuf library:

Reviewer: > could be in the application or in the python-tuf library

Isn't this a decision that we should make and give the rationale for in this design document?

Author: I think that we should include almost nothing like this in python-tuf as a starting point: I don't think we know e.g. how applications want to add keys (_multiple keys at once? are they always going to do  some other edits at the same time?_), and shouldn't try to provide solutions before we do... I'm just saying that if this turns out to be something repositories do a lot, we could later offer it as a helper function

Reviewer: Thanks for the explanation, it makes a lot of sense! Would you mind trying to
1. emphasise on the process-like nature of this design proposal for python-tuf (*start with almost nothing, then aggregate common functionality from applications as they are implemented*) possibly in the beginning of the "Design" section, and 
2. rephrase this paragraph to show your preference in this example (*key-adding code for now left to application as we don't know common usage pattern... multiple keys at once?, other edits at the same time?, etc.)*

?

Reviewer 2: Agree with the first suggestion.
To emphasize the first point will give more insight into the process of the design document creation which I think will be a valuable context.

```python
with repository.edit(“targets”) as targets:
    # adds a key for role1 (as an example, arbitrary edits are allowed)
    targets.add_key(“role1”, key)
```

This code loads current targets metadata for editing, adds the key to a role,
and handles version and expiry bumps before persisting the new targets version.
The reason for the context manager style is that it manages two things
simultaneously:
* Hides the complexity of loading and persisting metadata, and updating expiry
  and versions from the editing code (by putting it in the repository

>   implementation – which may still be provided by the application)

Reviewer: If I understand your design correctly, this  **must** (not may) be provided by the application, does it?

Author: yes that would be better phrasing. I've gone back and forth between mentioning that we could ship implementations in python-tuf source code: but as I think those would be examples and not library code, your phrasing is better

* Still allows completely arbitrary edits on the metadata in question: now the
  library does not need to anticipate what application wants to do and on the
  other hand library can still provide e.g. snapshot functionality without
  knowing about the application decisions mentioned in previous point.

Other designs do not seem to manage both of these.

> 

Reviewer: By reading the rendered markdown only, I wasn't sure whether you are describing two alternative designs in these diagrams, or different perspectives on the same design. This was reinforced by the preceding sentence "Other designs...".

Maybe adding a title, a la "How components are used" would do the trick.

![Design: How components are used](repository-library-design-usage.jpg)

The core idea here is that because editing is ergonomic enough, when new
functionality (like “developer uploads new targets”) is added, _it can be added
at any level_: the application might add a `handle_new_target_files()` method
that adds a bunch of targets into the metadata, but one of the previous layers
could offer that as a helper function as well: code in both cases would look

> similar as it would use the common editing interface.

Reviewer: Shouldn't the design document make the decision on what layer such functionality should be implemented? Or at least explain based on what factors the decision should be made?

Author: I'm proposing we start with the absolute minimum in the library: possibly just _snapshot_ and _sign_ implemented in the library itself. So the application is left with it's own design choices for the rest  (should `MyDerivedRepository` offer helper functions or should the application code just use the edit() contextmanager itself -- the proof of concept goes with the latter choice)... 

Nothing prevents us from moving some functionality into the library if that seems to make sense for multiple users but I don't see a reason to do that without seeing some implementations.

Reviewer: Again, with your explanation here the document makes more sense. (cf. my [reply above](https://github.com/theupdateframework/python-tuf/pull/1693/files/bcab2e96b03abd28aa180c7ad51b90e000fbc1dd#r760910439))

There are a few additional items worth mentioning:
* Private key management: the Repository API should come with a “keyring
  abstraction” -- a way for the application to provide roles’ private keys for
  the Repository to use. Some implementations could be provided as well.
* Validating repository state: the design is very much focused on enabling
  efficient editing of individual metadata. Implementations are also likely to
  be interested in validating (after some edits) that the repository is correct
  according to client workflow and that it contains the expected changes. The
  Repository API should provide some validation, but we should recognise that
  validation may be implementation specific.
* Improved metadata editing: There are a small number of improvements that
  could be made to metadata editing. These do not necessarily need to be part
  of the repository API: they could be part of Metadata API as well

It would make sense for python-tuf to ship with at least one concrete

> Repository implementation: possibly a repo.py look alike. This implementation

Reviewer 3: It might be nice to have this example concrete implementation support a CLI for demos and such.

Author: Yes. I didn't want to bloat these docs but I see a few potential implementations:
* some sort of CLI tool, to be able to generate and modify metadata for demo and testing purposes
* an example repository server (a "mini-warehouse"): a focused and opinionated repository example that could maybe include a custom web api for uploading new targets and a tiny developer tool to do the upload. It should be possible to make this into an example that runs out-of-the-box

Both of these could live inside or outside the repo... I'm not against the first one at all but I kind of feel the second one is even more important. The focus on CLI tools in TUF community (where every project seems to be a CLI tool) is not just a good thing IMO: it creates this impression that production repos can be shell scripts calling CLI tools**... which I don't think is such a great idea.

** _yes, I realize I've written exactly this myself: don't do as I do, do as I say :grimacing:_

| Concept | repository-editor-for-tuf implementation |
|-|-|
| Repository API | [librepo/repo.py](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/librepo/repo.py), [librepo/keys.py](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/librepo/repo.py)  |
| Example of repository implementation | [git_repo.py](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/git_repo.py) |
|Application code | [cli.py (command line app)](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/cli.py), [keys_impl.py (keyring implementation)](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/keys_impl.py) |
| Repository validation | [verifier.py (very rough, not intended for python-tuf)](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/verifier.py)
| Improved Metadata editing | [helpers.py](https://github.com/vmware-labs/repository-editor-for-tuf/blob/main/tufrepo/helpers.py) 

### Repository API

Repository itself is a minimal abstract class: The value of this class is in
defining the abstract method signatures (most importantly `_load`, `_save()`,

> `edit()`) that enable ergonomic metadata editing. The Repository class in this

Reviewer: Can `_load` and `_save` be seen as pre- and post-`edit` hooks? If that is correct, what is the benefit of prescribing hooks, if the thing that's being hooked into needs to be implemented by the users anyways? (apologies for not remembering/re-looking-up your implementation)

Author: yes they mostly can. `_load()` and `_save()` could be left as implementation details of the `edit()` implementation... but at least currently I've implemented `sign()` as 
```python
        md = self._load(role)
        self._save(role, md, clear_sigs=False)
```
in the abstract class -- I can't call `edit()` as it could modify the version, expiry, etc...  I think this is the main reason _load/_save are defined.

You may be right that we should figure out some alternative solution for sign(), and the _load/_save should be implementation details of `edit()` implementation

Reviewer: Would it be possible to explain the relationship/dependencies between abstract interfaces `_load/_save/edit` and concrete implementations `sign/snapshot` a little bit more in this paragraph?
IIUC, in order to use the concrete implementation, the application is required to first implement the abstract interfaces, right? I think that's worth mentioning, to better understand the whole "what's provided by python-tuf and what isn't"-thing. :)

Author: This makes sense, I'll add a separate section for this

* Private keys can be stored in a file or in environment variables (for CI use)

Note that GitRepository implementation is significantly larger than the
Repository interface -- but all of the complexity in GitRepository is really
related to the design decisions made there.

_Prototype status: The GitRepository example exists in git_repo.py._

### Validating repository state

This is mostly undesigned but something built on top of TrustedMetadataSet
(currently ngclient component) might work as a way to easily check specific
aspects like:
* Is top-level metadata valid according to client workflow
* Is a role included in the snapshot and the delegation tree

It’s likely that different implementations will have different needs though: a
command line app for small repos might want to validate loading all metadata
into memory, but a server application hosting tens of thousands of pieces of
metadata is unlikely to do so.

_Prototype status: A very rough implementation exists in verifier.py : this is
unlikely to be very useful_

### Improved metadata editing

Currently the identified improvement areas are:
* Metadata initialization: this could potentially be improved by adding

>   default argument values to Metadata API constructors 

Reviewer: I see potential problems for our current deserialization implementation when providing default arguments, but we can discuss this separately.

Author: let's do that in #1459 (see https://github.com/theupdateframework/python-tuf/issues/1459#issuecomment-953729281)

Possibly some `from_dict()` implementations would have to be extra careful not to pass deserialized None/null values to a constructor.

Repository implementation: possibly a repo.py look alike. This implementation
should not be part of the library but an example.

### What a concrete Repository implementation must implement

The current prototype API requires five methods that take care of access to
metadata storage, expiry updates, version updates and signing.

> * **keyring()**: A property that returns the private key Mapping that should be

Reviewer: ```suggestion
* **keyring()**: A property that returns the private key mapping that should be
```
?

  used for signing

* **_load()**: Loads metadata from storage or cache. Is used by edit() and

> sign()

Reviewer: You reader only learns about `sign()` in the next section in the context of your POC implementation. Maybe this is confusing here in the more generic part of the document? Any reason to not move this entire list down to the Details section?

Author: Hmm, this is all true, will change as suggested

> * **_save()**: signs and persits metadata in cache/storage. Is used by edit()

Reviewer: ```suggestion
* **_save()**: Signs and persists metadata in cache/storage. Is used by edit()
```

and sign()

> * **edit()**: The ContextManager that enables ergonomic metadata metadata

Reviewer: ```suggestion
* **edit()**: The ContextManager that enables ergonomic metadata
```
